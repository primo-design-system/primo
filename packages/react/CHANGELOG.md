# Change Log

All notable changes to this project will be documented in this file. See
[Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.5](https://github.com/primodesignsystem/primo/compare/@primo/react@1.0.4...@primo/react@1.0.5) (2023-01-10)

**Note:** Version bump only for package @primo/react

# Change Log

All notable changes to this project will be documented in this file. See
[Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.4](https://github.com/primodesignsystem/primo/compare/@primo/react@1.0.2...@primo/react@1.0.4) (2023-01-09)

**Note:** Version bump only for package @primo/react

## @primo/react [1.0.2](https://github.com/primo-design-system/primo/compare/@primo/react@1.0.1...@primo/react@1.0.2) (2022-12-12)

### Dependencies

- **@primo/tokens:** upgraded to 1.0.1
- **@primo/styles:** upgraded to 1.1.0
- **@primo/components:** upgraded to 2.0.0

## @primo/react [1.0.1](https://github.com/primo-design-system/primo/compare/@primo/react@1.0.0...@primo/react@1.0.1) (2022-11-19)

### Dependencies

- **@primo/styles:** upgraded to 1.0.1
- **@primo/components:** upgraded to 1.0.1

# @primo/react 1.0.0 (2022-11-18)

### Features

- new button changes + playrom and new tests
  ([c1da249](https://github.com/primo-design-system/primo/commit/c1da24929b594f069e18bc40003041e37f4b6811))
- updating docs and tests
  ([0f95c7c](https://github.com/primo-design-system/primo/commit/0f95c7ca2bd271f25333f017d7b5e770d4c43462))
- **component:** build scripts for button component and linting
  ([ba76456](https://github.com/primo-design-system/primo/commit/ba76456f523a53d09c0605615ebf9f7751f129a7))
- **react:** new ButtonGroup component, setup and tests
  ([8685ea5](https://github.com/primo-design-system/primo/commit/8685ea55fd7d6dba68eb72240bdbcaece1bdaf43))
- **tokens:** including new token values
  ([10f1161](https://github.com/primo-design-system/primo/commit/10f11615e87e00bcc691c18ccd04913c1bec8362))
- **tokens:** new colour and font tokens
  ([38efa21](https://github.com/primo-design-system/primo/commit/38efa21d6e0c487f10e5aacf6f0a030d3170dcf1))

### Dependencies

- **@primo/tokens:** upgraded to 1.0.0
- **@primo/styles:** upgraded to 1.0.0
- **@primo/components:** upgraded to 1.0.0
