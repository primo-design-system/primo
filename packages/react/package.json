{
  "name": "@primo/react",
  "version": "1.0.2",
  "description": "> TODO: description",
  "author": "Richard McCartney <richard@mccartney.io>",
  "homepage": "https://github.com/primodesignsystem/primo#readme",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/index.es.js",
  "style": "dist/index.css",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c",
    "build:playroom": "playroom build",
    "build:storybook": "build-storybook --docs --quiet",
    "build:storybook:quiet": "yarn storybook --quiet",
    "build:watch": "rollup -c -w",
    "lint": "yarn lint:es && yarn lint:ts",
    "lint:es": "eslint src/**/*.ts[x]",
    "lint:es:fix": "yarn lint:es --fix",
    "lint:ts": "tslint -c tslint.json 'src/**/*.{ts,tsx}'",
    "lint:ts:fix": "yarn lint:ts --fix",
    "start:playroom": "playroom start",
    "start:storybook": "start-storybook -p 6006 -s public",
    "test": "jest",
    "test:update-snapshots": "yarn test -u"
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org",
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/primodesignsystem/primo.git"
  },
  "bugs": {
    "url": "https://github.com/primodesignsystem/primo/issues"
  },
  "devDependencies": {
    "@babel/core": "^7.11.5",
    "@primo/tokens": "1.0.1",
    "@primo/styles": "1.1.0",
    "@primo/components": "2.0.0",
    "@rollup/plugin-commonjs": "^23.0.2",
    "@rollup/plugin-node-resolve": "^15.0.1",
    "@rollup/plugin-typescript": "^9.0.2",
    "@storybook/addon-a11y": "^6.2.9",
    "@storybook/addon-essentials": "^6.2.9",
    "@storybook/addon-links": "^6.2.9",
    "@storybook/addon-postcss": "^2.0.0",
    "@storybook/react": "^6.5.12",
    "@storybook/theming": "^6.2.9",
    "@testing-library/jest-dom": "^5.11.4",
    "@testing-library/react": "^11.0.4",
    "@types/classnames": "^2.2.10",
    "@types/jest": "^26.0.14",
    "@types/jest-axe": "^3.5.5",
    "@types/node": "^12.12.38",
    "@types/prop-types": "^15.7.3",
    "@types/react": "^16.9.49",
    "@types/react-dom": "^16.9.8",
    "@types/storybook__react": "^5.2.1",
    "@types/testing-library__jest-dom": "^5.9.2",
    "@types/testing-library__react": "^10.2.0",
    "@typescript-eslint/eslint-plugin": "^4.0.0",
    "@typescript-eslint/parser": "^4.0.0",
    "@whitespace/storybook-addon-html": "^5.0.0",
    "babel-eslint": "^10.0.0",
    "babel-loader": "^8.1.0",
    "cheerio": "^1.0.0-rc.3",
    "css-loader": "^6.7.1",
    "file-loader": "6.2.0",
    "jest": "^26.4.2",
    "jest-axe": "^6.0.0",
    "playroom": "0.28.0",
    "postcss": "^8.2.4",
    "postcss-import": "^12.0.1",
    "postcss-preset-env": "^6.7.0",
    "prettier": "^1.19.1",
    "prop-types": "^15.7.2",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-is": "^16.13.1",
    "rollup": "^2.26.9",
    "rollup-plugin-peer-deps-external": "^2.2.3",
    "rollup-plugin-postcss": "^3.1.8",
    "style-loader": "^3.3.1",
    "ts-jest": "^26.4.0",
    "tslib": "^2.0.1",
    "tslint": "^6.1.3",
    "typescript": "^4.0.2",
    "url-loader": "^4.1.1",
    "webpack": "^5.2.0"
  }
}
