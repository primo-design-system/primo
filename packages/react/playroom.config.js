const exampleCode = `
  <Button>Hello World</Button>
`;

module.exports = {
  components: './src/components/index.ts',
  outputPath: './dist/playroom',

  // Optional:
  title: 'Primo Design System',
  widths: [320, 768, 1024],
  port: 9000,
  openBrowser: true,
  paramType: 'search', // default is 'hash'
  frameComponent: './.playroom/FrameComponent.tsx',
  exampleCode: exampleCode,
  baseUrl: '/playroom/',
  webpackConfig: () => ({
    module: {
      rules: [
        {
          test: /\.tsx?$/,
          include: __dirname,
          use: {
            loader: 'babel-loader',
            options: {
              presets: ['@babel/preset-env', '@babel/preset-typescript', '@babel/preset-react'],
            },
          },
        },
        {
          test: /\.css$/,
          exclude: /node_modules\/(?!(@primo)\/).*/,
          use: [require.resolve('style-loader'), require.resolve('css-loader')],
        },
      ],
    },
    resolve: {
      extensions: ['.js', '.ts', '.tsx', '.css', '.pcss'],
      modules: ['node_modules'],
    },
  }),
  iframeSandbox: 'allow-scripts',
};
