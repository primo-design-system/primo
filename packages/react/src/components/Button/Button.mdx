<!-- Button.stories.mdx --->

import {
  Story,
  ArgsTable,
  Source,
  Preview,
  PRIMARY_STORY,
} from '@storybook/addon-docs/blocks';
import Button from './Button';

# Button

[Source code](https://bitbucket.org/lloyds-digital-market/cardano/src/main/packages/react/src/components/Button)
&nbsp;|&nbsp; [Usage guidelines](#) &nbsp;|&nbsp; [Accessibility](#)

## Table of Contents

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Overview](#overview)
- [Component API](#component-api)
- [Feedback](#feedback)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

Buttons are clickable elements that are used to trigger actions.

<Preview>
  <Story id="components-button--default" />
</Preview>

## Component API

### Links props

<ArgsTable exclude={['name', 'type', 'value', 'endIcon', 'startIcon']} />

### Button props

<ArgsTable exclude={['download', 'href', 'hrefLang', 'rel', 'target']} />

## Stories

### Primary Buttons

<Preview>
  <Story id="components-button--primary" />
</Preview>

### Negative Buttons

Negative Buttons have three styles, determining which negative button style to
use will depend on the level of emphasis you want to give to the negative action

<Preview>
  <Story id="components-button--negative" />
</Preview>

### Full Width Buttons

<Preview>
  <Story id="components-button--full-width" />
</Preview>

### Small Buttons

<Preview>
  <Story id="components-button--small" />
</Preview>

### With Icon Buttons

<Preview>
  <Story id="components-button--with-icon" />
</Preview>

### Only Icon Buttons

<Preview>
  <Story id="components-button--only-icon" />
</Preview>

### Feedback

Please help us improve the component or documentation, provide feedback and come
with any questions in Slack or updating this file on
[BitBucket](https://bitbucket.org/lloyds-digital-market/cardano/src/main/packages/react/src/components/Button/Button.mdx)
