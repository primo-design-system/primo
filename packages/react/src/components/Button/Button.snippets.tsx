export const ButtonSnippets = [
  {
    group: 'Button',
    name: 'Default',
    code: `
      <Button>
        Button
      </Button>
    `,
  },
  {
    group: 'Button',
    name: 'Default Full Width',
    code: `
      <Button isFullWidth>
        Button Full Width
      </Button>
    `,
  },
  {
    group: 'Button',
    name: 'Default Link',
    code: `
      <Button href="#">
        Button Link
      </Button>
    `,
  },
  {
    group: 'Button',
    name: 'Default Disabled',
    code: `
      <Button isDisabled>
        Button Disabled
      </Button>
    `,
  },
  {
    group: 'Button',
    name: 'Primary',
    code: `
      <Button variant="primary">
        Primary Button
      </Button>
    `,
  },

  {
    group: 'Button',
    name: 'Negative',
    code: `
      <Button variant="negative">
        Negative Button
      </Button>
    `,
  },
  {
    group: 'Button',
    name: 'Negative Outline',
    code: `
      <Button variant="negative" fill="outline">
        Negative Button
      </Button>
    `,
  },
];
