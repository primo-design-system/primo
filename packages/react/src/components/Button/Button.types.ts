import React, { MouseEvent, ReactNode } from 'react';
// import { IconTypes } from '../Icon/Icon.types';

export type AnchorAttributes = React.AnchorHTMLAttributes<HTMLAnchorElement>;
export type ButtonAttributes = React.ButtonHTMLAttributes<HTMLButtonElement>;

export type RefType = React.Ref<HTMLButtonElement | HTMLAnchorElement>;

export type ButtonFill = 'default' | 'ghost' | 'outline';
export type ButtonTarget = '_blank' | '_parent' | '_self' | '_top';
export type ButtonType = 'button' | 'reset' | 'submit';
export type ButtonVariant = 'default' | 'primary' | 'secondary' | 'tertiary' | 'positive' | 'negative';

interface SharedButtonProps {
  /** Specify the content of your Button */
  children?: ReactNode;

  /** Optional prop to specify classNames onto the Button */
  className?: string;

  /** Optional prop to specify the ID of the Button */
  id?: string;

  /** Optionally specify Button onClick function */
  onClick?: (e: MouseEvent<HTMLElement>) => void;

  /** Specify whether the Button is disables, or not */
  isDisabled?: boolean;

  /** Specify whether the Button is full width or not */
  isFullWidth?: boolean;

  /** Specify whether the Button is sized small or not */
  isSmall?: boolean;

  /** Optional prop to specify the icon label of the Button */
  // iconLabel?: string;

  // /** Optional prop to specify an Icon in the end position of the Button */
  // endIcon?: IconTypes;

  // /** Optional prop to specify an Icon in the start position of the Button */
  // startIcon?: IconTypes;

  /**
   * Optional prop to specify the fill style of the Button
   * @default 'default'
   */
  fill?: ButtonFill;

  /** Optional prop to specify the tabIndex of the Button */
  tabIndex?: number;

  /** Optional prop to specify the ID used for testing */
  testId?: string;

  /**
   * Specify the variant of the Button
   * @default 'default'
   */
  variant?: ButtonVariant;

  /** Optional property to provide component Ref */
  ref?: RefType;
}

export interface ButtonElementProps extends SharedButtonProps {
  /** Optional prop to specify the name of the Button */
  name?: string;

  /**
   * Optional prop to specify the type of the Button
   * @default 'button'
   */
  type?: ButtonType;

  /** Optional prop to specify the value of the Button */
  value?: string;

  // /** Only applies to <a />s */
  download?: never;
  href?: never;
  hrefLang?: never;
  rel?: never;
  target?: never;
}

export interface LinkElementProps extends SharedButtonProps {
  /** Optional prop to prompts a user to save the linked URL instead of navigating to it */
  download?: string;

  /** Optionally specify an href for your Button to become an <a> element */
  href?: string;

  /** Optionally specify the language of the linked URL */
  hrefLang?: string;

  /**
   * Relationship between the Button href URL and the current page. See
   * [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel).
   */
  rel?: string;

  /** Optional prop to to display where the linked URL will show */
  target?: ButtonTarget;

  /** Only applies to <button />s */
  name?: never;
  type?: never;
  value?: never;
  endIcon?: never;
  startIcon?: never;
}

export type ButtonProps = ButtonElementProps | LinkElementProps;

export type PolymorphicButton = (
  props: LinkElementProps | ButtonElementProps
) => JSX.Element;
