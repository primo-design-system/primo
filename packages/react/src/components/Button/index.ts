export { default } from './Button';
export type {
  ButtonProps,
  LinkElementProps,
  ButtonElementProps,
} from './Button.types';
