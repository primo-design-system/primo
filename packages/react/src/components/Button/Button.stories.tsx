// Button.stories.js
import React from 'react';
import { Meta, Story } from '@storybook/react';
import Button, { ButtonProps } from '.';
import mdx from './Button.mdx';

export default {
  title: 'Components/Button',
  args: { children: 'Button' },
  parameters: {
    docs: {
      page: mdx,
    },
  },
  component: Button,
} as Meta;

export const Default: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} />
);

export const Primary: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} variant="primary" />
);

export const Secondary: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} variant="secondary" />
);

export const Tertiary: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} variant="tertiary" />
);

export const Positive: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} variant="positive" />
);

export const Negative: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} variant="negative" />
);

export const FullWidth: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} isFullWidth />
);

export const Small: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} size="small" />
);

export const Large: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} size="large" />
);

export const Pill: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} shape="pill" />
);

export const Round: Story<ButtonProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
  <Button {...args} shape="round" />
);

// export const WithIcon: Story<ButtonElementProps> = (args: JSX.IntrinsicAttributes & React.PropsWithChildren<ButtonProps>) => (
//   <Button {...args} startIcon="copy" />
// );

// export const OnlyIcon: Story<ButtonProps> = () => (
//   <Button startIcon="copy" iconLabel="copy" />
// );
