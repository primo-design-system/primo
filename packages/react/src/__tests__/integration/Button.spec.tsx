// src/__tests__/integration/Button.spec.tsx

import * as React from 'react';
import { render, RenderResult } from '@testing-library/react';

import Button from '../../components/Button'

let documentBody: RenderResult;

describe('<Button />', () => {
  beforeEach(() => {
    documentBody = render(<Button>Click me!</Button>);
  });
  it(': shows "Click me!" text', () => {
    expect(documentBody.getByText('Click me!')).toBeInTheDocument();
  });

  it(': matches snapshot', () => {
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });
});
