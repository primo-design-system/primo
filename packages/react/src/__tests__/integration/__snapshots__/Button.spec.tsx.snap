// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> : matches snapshot 1`] = `
<body>
  <div>
    <button
      class="primo-Button"
      type="button"
    >
      Click me!
    </button>
  </div>
</body>
`;
